Fix: Azure GPT-5 incorrectly routed to O-series config (temperature parameter unsupported)

xingyaoww opened the pull request:
## Problem

Azure GPT-5 users were experiencing this error when using the Responses API with the `temperature` parameter:

```python
litellm.UnsupportedParamsError: LlmProviders.AZURE does not support parameters: {'temperature': 1.0}, for model=gpt-5
```

However:
- ✅ OpenAI GPT-5 with temperature worked fine
- ✅ Azure officially supports temperature parameter (per [Azure docs](https://learn.microsoft.com/en-us/azure/ai-foundry/openai/reference-preview-latest#request-body-14))
- ❌ Only Azure GPT-5 threw the error

Source issue: fix https://github.com/OpenHands/software-agent-sdk/issues/986


## Root Cause

The bug was in the routing logic in `ProviderConfigManager.get_provider_responses_api_config()` at **line 7400** of `litellm/utils.py`.

**The Issue:**
```python
# Old (broken) code
if model and ("o_series" in model.lower() or supports_reasoning(model)):
    return litellm.AzureOpenAIOSeriesResponsesAPIConfig()
```

Since `supports_reasoning("gpt-5")` returns `True`, GPT-5 was incorrectly routed to `AzureOpenAIOSeriesResponsesAPIConfig`, which explicitly removes `temperature` from supported parameters (correct for O-series models like o1/o3, but wrong for GPT-5).

**Why it matters:**
- GPT-5 supports reasoning (extended thinking) but is **NOT** an O-series model
- GPT-5 **DOES** support the temperature parameter
- O-series models (o1, o3) are a specific family with different parameter restrictions

## Solution

Added explicit logic to exclude GPT-5 from O-series routing:

```python
# New (fixed) code
is_gpt5 = model and "gpt-5" in model.lower()
is_o_series = model and ("o_series" in model.lower() or (supports_reasoning(model) and not is_gpt5))

if is_o_series:
    return litellm.AzureOpenAIOSeriesResponsesAPIConfig()
else:
    return litellm.AzureOpenAIResponsesAPIConfig()
```

**Result:**
- ✅ Azure GPT-5 → `AzureOpenAIResponsesAPIConfig` (supports temperature)
- ✅ Azure O1/O3 → `AzureOpenAIOSeriesResponsesAPIConfig` (no temperature)
- ✅ OpenAI GPT-5 → `OpenAIResponsesAPIConfig` (supports temperature)

## Testing

Added comprehensive unit tests in `tests/test_gpt5_azure_temperature_support.py`:

1. ✅ `test_azure_gpt5_supports_temperature()` - Verifies Azure GPT-5 uses correct config
2. ✅ `test_azure_o_series_does_not_support_temperature()` - Verifies O-series routing still works
3. ✅ `test_openai_gpt5_supports_temperature()` - Verifies OpenAI GPT-5 support
4. ✅ `test_azure_gpt5_variants_support_temperature()` - Tests various GPT-5 name formats

All tests pass! Run with:
```bash
pytest tests/test_gpt5_azure_temperature_support.py -v
```

## Manual Verification

```python
from litellm.utils import ProviderConfigManager
from litellm.types.utils import LlmProviders

# Test Azure GPT-5
config = ProviderConfigManager.get_provider_responses_api_config(
    provider=LlmProviders.AZURE,
    model="gpt-5"
)
print(f"Config: {type(config).__name__}")  # AzureOpenAIResponsesAPIConfig ✅
print(f"Supports temperature: {'temperature' in config.get_supported_openai_params('gpt-5')}")  # True ✅
```

## Files Changed

- `litellm/utils.py` - Fixed routing logic (6 lines changed)
- `tests/test_gpt5_azure_temperature_support.py` - Added comprehensive tests (80 lines)

## Impact

- **Backwards Compatible**: ✅ No breaking changes
- **O-series models**: ✅ Still work correctly (no temperature)
- **GPT-5 models**: ✅ Now work correctly (with temperature)
- **Other models**: ✅ Unaffected

## Checklist

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Tests added for the fix
- [x] Manual verification completed
- [x] No breaking changes
- [x] Documentation in code comments

@xingyaoww can click here to [continue refining the PR](https://app.all-hands.dev/conversations/f876aee5bab7450e9abcfb999450fc89)

openhands-agent committed:
Fix Azure GPT-5 incorrectly routing to O-series config

GPT-5 models support reasoning but are NOT O-series models and DO support
temperature parameter. The previous routing logic in get_provider_responses_api_config()
was incorrectly sending Azure GPT-5 requests to AzureOpenAIOSeriesResponsesAPIConfig
which removes temperature from supported params.

This fix explicitly excludes GPT-5 models from O-series routing, ensuring they
use the standard AzureOpenAIResponsesAPIConfig which properly supports temperature.

Fixes: Azure GPT-5 throwing UnsupportedParamsError for temperature parameter
Tested: Added comprehensive unit tests for GPT-5 and O-series routing

xingyaoww committed:
Apply suggestion from @xingyaoww

xingyaoww commented on tests/test_gpt5_azure_temperature_support.py:
```diff
@@ -0,0 +1,79 @@
+"""
+Test that Azure GPT-5 models support temperature parameter in Responses API.
+
+This test verifies the fix for: https://github.com/BerriAI/litellm/issues/XXXXX
```
```suggestion
```

xingyaoww committed:
Apply suggestion from @xingyaoww

krrishdholakia commented on litellm/utils.py:
```diff
@@ -7397,7 +7397,11 @@ def get_provider_responses_api_config(
-            if model and ("o_series" in model.lower() or supports_reasoning(model)):
+            # Note: GPT-5 supports reasoning but is NOT an O-series model and supports temperature
+            is_gpt5 = model and "gpt-5" in model.lower()
+            is_o_series = model and ("o_series" in model.lower() or (supports_reasoning(model) and not is_gpt5))
```
why not just avoid routing here if it has "gpt" in model name - it's broader and will probably cover future openai models better @xingyaoww

openhands-agent committed:
Improve Azure routing logic to use broader 'gpt' check for temperature support

Based on feedback from @krrishdholakia, updated the routing logic to check
for 'gpt' in model name instead of specifically 'gpt-5'. This approach is:

- More future-proof: covers all GPT models (gpt-3.5, gpt-4, gpt-5, future models)
- Simpler: single check for all GPT variants
- More maintainable: won't need updates for each new GPT model

Changes:
- litellm/utils.py: Changed from is_gpt5 to is_gpt_model check
- tests: Added comprehensive test for all GPT model variants (gpt-3.5 through gpt-5)

All tests pass:
- GPT models (gpt-3.5-turbo, gpt-4, gpt-4o, gpt-5) -> AzureOpenAIResponsesAPIConfig (supports temperature)
- O-series models (o1, o3) -> AzureOpenAIOSeriesResponsesAPIConfig (no temperature)

Co-authored-by: openhands <openhands@all-hands.dev>
